### PR TITLE
Refactor Lighting Module Configuration and Optimizer Set

### PR DESCRIPTION
**Pull request by TonyCongqianWang**
Two major changes:

1. Refactoring configs across the project to be more consistent with tyro:
* Using nested configs for lightning module.
* Modernized `serialize.py` cli with tyro.

2. Refactoring lightning modules Optimizer setup. Added wrapper classes for optimizers so that swapping to different optimizers creates less friction.

* Adding schedulefree as possible optimizer.
* Default behavior is not changed

* Using ranger21 from pip instead of custom file. Note: after `pip install ranger21` using `diff /usr/local/lib/python3.12/dist-packages/ranger21/ranger21.py ranger21.py` reveals that reveals that the `ranger21.py` file in master is identical to the one installed by pip

This might be useful for anyone that wants to try schedulefree. Also it opens the path to add other optimizers in the future. Maybe the best overall recipe uses different optimizers between stages

The reason why I think it us useful to merge now even though schedulefree hasn't been successful so far is that it removes friction in the future (especially in terms of merge conflics) when we want to test other optimizers.

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -5,11 +5,12 @@
 
 from data_loader.config import DataloaderSkipConfig
 from model.config import LossParams, ModelConfig
+from model.optimizers.optimizer_config import OptimizerConfig
```
Comment (xu-shawn) [inline config.py]: could you export the classes through an `__init__.py` file under model/optimizers?

File: train.py [hunk 2]
```diff
@@ -166,10 +166,14 @@ def main():
...
             feature_name=feature_name,
             loss_params=loss_params,
+            optimizer_name=args.optimizer_name,
             max_epoch=max_epoch,
             num_batches_per_epoch=num_batches_per_epoch,
             gamma=args.gamma,
+            ft_weight_decay=args.ft_weight_decay,
+            dense_weight_decay=args.dense_weight_decay,
             lr=args.lr,
+            warmup_steps=args.warmup_steps,
```
Comment (xu-shawn) [inline train.py]: if we were to add these parameters, it would be a good idea to group them into OptimizerConfig and do something like this to extract argparse results:
https://github.com/official-stockfish/nnue-pytorch/blob/e47dde7f9746c7522b58bb5c09016653888f4f05/data_loader/config.py#L27-L38

File: config.py (hunk 1 above)
Comment (TonyCongqianWang) [reply config.py]: I could. But do we want them to be "private" or public?

File: train.py (hunk 2 above)
Comment (TonyCongqianWang) [reply train.py]: I also thought about it... We could do composition instead of inheritance with the parameters... Would be weird if it's only here and inheritance everywhere else. I'd suggest it to be somewhat consistent

Comment (TonyCongqianWang) [reply train.py]: I am not sure If I understand you correctly though.. can you give me an example of what you think would be good?

 I am using a dataclass `OptimizerConfig` and tyro to extract the parameters

File: config.py (hunk 1 above)
Comment (xu-shawn) [reply config.py]: imo it should make sense to export `OptimizerConfig`, `Ranger21Wrapper`, and `ScheduleFreeWrapper`

File: train.py (hunk 2 above)
Comment (xu-shawn) [reply train.py]: I'm suggesting something like this:

```py
@dataclass
class OptimizerConfig:
    ...

    def get_optimizer_config_from_args(args) -> "OptimizerConfig"
         ...
         return OptimizerConfig(...)
```

And refactoring the `NNUE` class to take in an `OptimizerConfig` in the constructor instead of various separate optimizer parameters

Comment (Disservin) [reply train.py]: get_dataloader_skip_config_from_args is actually not really needed anymore.. you can replace the current call side with just args, since it inherits from DataloaderSkipConfig

Comment (xu-shawn) [reply train.py]: ah I guess then just refactoring the NNUE class constructor should be good

Comment (Disservin) [reply train.py]: yeah though it feels not so clean when you pass the entire object and rely on type hints, so im nto sure what to do

Comment (TonyCongqianWang) [reply train.py]: The best solution would be to use nested configs (composition) instead of inheritance. Tyro can handle them very well.

Then you can indeed pass the optimizer config class which would the best.

But I would suggest doing all of the config refactoring in one PR

File: config.py (hunk 1 above)
Comment (TonyCongqianWang) [reply config.py]: Ok sure. Can do.

I'm gonna work on it tomorrow

File: train.py (hunk 2 above)
Comment (Disservin) [reply train.py]: yea see https://github.com/official-stockfish/nnue-pytorch/pull/425

File: config.py (hunk 1 above)
Comment (TonyCongqianWang) [reply config.py]: Done, please review :)

File: train.py (hunk 2 above)
Comment (TonyCongqianWang) [reply train.py]: Done, please review. Probably gonna throw out argparse at serialize.py as well but I just wanted something worked before going to bed...

File: serialize.py [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
 import model as M
 
-
 def main():
-    parser = argparse.ArgumentParser(
-        description="Converts files between ckpt and nnue format."
+    class NoExitParser(argparse.ArgumentParser):
+        def error(self, message):
+            # Instead of exiting, we just record that it failed or ignore it
+            print(f"Argparse ignored: {message}")
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply serialize.py]: I will let Gemini write me a tyro class to get rid of the argparse hack

File: train.py [hunk 4]
```diff
@@ -203,14 +191,12 @@ def main():
...
-        nnue.num_batches_per_epoch = num_batches_per_epoch
         # we can set the following here just like that because when resuming
         # from .pt the optimizer is only created after the training is started
-        nnue.gamma = args.gamma
-        nnue.lr = args.lr
-        nnue.compile_backend = args.compile_backend
+        nnue.loss_params = loss_params
+        nnue.max_epoch = max_epoch
+        nnue.num_batches_per_epoch=args.num_batches_per_epoch
+        args.nnue_lightning_config = args.nnue_lightning_config
```
Comment (Disservin) [inline train.py]: args.nnue_lightning_config = args.nnue_lightning_config ?

File: serialize.py [hunk 5]
```diff
@@ -95,65 +85,65 @@ def main():
     elif args.source.endswith(".nnue"):
         with open(args.source, "rb") as f:
             nnue = M.NNUE(
-                feature_name,
-                M.ModelConfig.get_model_config(args),
-                M.QuantizationConfig(),
+                config=nnue_lightning_config,
+                quantize_config=M.QuantizationConfig(),
             )
```
Comment (Disservin) [inline serialize.py:90]: what happened to feature_name?

File: train.py (hunk 4 above)
Comment (TonyCongqianWang) [reply train.py]: Oops...

Will fix! Thanks for the review

File: serialize.py (hunk 5 above)
Comment (TonyCongqianWang) [reply serialize.py:90]: I is now part of the model config

Comment (Disservin) [reply serialize.py:90]: nvm

File: train.py [hunk 6]
```diff
@@ -173,26 +169,18 @@ def main():
...
 
-    feature_name = args.features
+    feature_name = args.nnue_lightning_config.features
 
     print("Loss parameters:")
     print(loss_params)
 
-    num_batches_per_epoch=max(
-                1, args.epoch_size // global_batch_size_requested
-            )
```
Comment (Disservin) [inline train.py:183]: is this correct that this is gone now?

Comment (Disservin) [reply train.py:183]: ah i see it's a property okay

Comment (TonyCongqianWang) [reply train.py:183]: Yes!

It is moved inside the training config as property as it can be calculated from batch size and epoch size, both part of training config

File: model/optimizers/ranger21_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    num_batches_per_epoch: int = 100_000_000 // 16384
+    """Number of batches per epoch, used by Ranger21 for scheduling."""
+
+class Ranger21Wrapper:
+    def __init__(
+        self,
+        config,
+        max_epoch,
+        num_batches_per_epoch,
```
Comment (xu-shawn) [inline model/optimizers/ranger21_wrapper.py:23]: is this argument ever used?

File: model/optimizers/ranger21_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    ):
+        self.max_epoch = max_epoch
+        self.num_batches_per_epoch = config.num_batches_per_epoch
```
Comment (xu-shawn) [inline model/optimizers/ranger21_wrapper.py]: if I understand correctly, `num_batches_per_epoch` is now calculated in `TrainingConfig` and passed via function/constructor parameters, but here, `config.num_batches_per_epoch` is the default value from `OptimizerConfig`, which is inherited from `Ranger21Config` and never modified

File: model/optimizers/ranger21_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    _ranger21_import_error = False
+except ImportError:
+    _ranger21_import_error = True
+
+@dataclass
+class Ranger21Config:
+    gamma: float = 0.992
+    """Multiplicative factor applied to the learning rate after every epoch."""
+
+    num_batches_per_epoch: int = 100_000_000 // 16384
```
Comment (xu-shawn) [inline model/optimizers/ranger21_wrapper.py]: should this be removed?

File: model/optimizers/ranger21_wrapper.py (hunk 7 above)
Comment (TonyCongqianWang) [reply model/optimizers/ranger21_wrapper.py:23]: Honestly I don't know.. we could look inside the ranger21 code.

As warmup and warm down are both disabled it might be unused.

Master branch does pass it regardless however

File: model/optimizers/ranger21_wrapper.py (hunk 8 above)
Comment (TonyCongqianWang) [reply model/optimizers/ranger21_wrapper.py]: Oops, this is definitely a mistake yes. Will fix

File: model/optimizers/ranger21_wrapper.py (hunk 7 above)
Comment (xu-shawn) [reply model/optimizers/ranger21_wrapper.py:23]: oh I meant that it's not used in the constructor. I think I elaborated on it more clearly in the review below, so I'll just resolve this

Comment (TonyCongqianWang) [reply model/optimizers/ranger21_wrapper.py:23]: Ahh I see haha. Misunderstanding.

But I did look at the code.. and it does seem like num_batches_per_epoch is essentially unused by ranger21 in our setup...

File: config.py [hunk 10]
```diff
@@ -82,10 +74,19 @@ class TrainingConfig(FeatureConfig):
...
-    model_config: OmitArgPrefixes[ModelConfig] = field(default_factory=ModelConfig)
+    nnue_lightning_config: OmitArgPrefixes[NNUELightningConfig] = field(
+        default_factory=NNUELightningConfig
+    )
 
-    loss_config: OmitArgPrefixes[LossParams] = field(default_factory=LossParams)
+    @property
+    def batch_size(self) -> int:
+        """Returns the validated batch size."""
+        return self.batch_size_raw if self.batch_size_raw > 0 else 16384
```
Comment (xu-shawn) [inline config.py]: is this check now unnecessary? batch_size_raw is now 16384 by default

File: train.py [hunk 11]
```diff
@@ -203,14 +191,12 @@ def main():
...
-        nnue.max_epoch = max_epoch
-        nnue.num_batches_per_epoch = num_batches_per_epoch
         # we can set the following here just like that because when resuming
         # from .pt the optimizer is only created after the training is started
-        nnue.gamma = args.gamma
-        nnue.lr = args.lr
-        nnue.compile_backend = args.compile_backend
+        nnue.loss_params = loss_params
+        nnue.max_epoch = max_epoch
+        nnue.num_batches_per_epoch=args.num_batches_per_epoch
```
Comment (xu-shawn) [inline train.py]: minor formatting inconsistency. did you run ruff format?

File: config.py (hunk 10 above)
Comment (TonyCongqianWang) [reply config.py]: That's true. Should we just throw an error if someone tries to pass something <=0?

File: train.py (hunk 11 above)
Comment (TonyCongqianWang) [reply train.py]: Oh lol didn't know about it. Relied on vcscode so far.... Which doesn't handle python super reliably 

And also the GitHub automated workflow of course. So it seems ruff didn't flag thig inconsistency. Good spot

File: config.py (hunk 10 above)
Comment (TonyCongqianWang) [reply config.py]: Turns out input validation with tyro is surprisingly difficult. There seems to be a bug or wrong documentation for the custom constructors.

Also surprisingly there is no equivalent to nargs='+' in tyro such that for positional arguments of type list, an empty list is always valid
